### PR TITLE
core: Add API to allow signal listeners to constrain geometry

### PR DESCRIPTION
**Pull request by smspillaz**
https://phabricator.endlessm.com/T15968

### CODE REVIEW

File: src/core/window.c [hunk 1]
```diff
@@ -3652,6 +3671,32 @@ meta_window_update_monitor (MetaWindow *window,
...
+                  window->currently_allocating_rect.width : width),
+        .height = (height < window->currently_allocating_rect.height ?
+                   window->currently_allocating_rect.height : height)
+    };
+
+    window->currently_allocating_rect = next;
+}
+
+void
+meta_window_get_minimum_size_allocation (MetaWindow *window,
```
Comment (cosimoc) [inline src/core/window.c]: This function is unused both in this patchset and the one for eos-desktop.

File: src/core/window.c [hunk 2]
```diff
@@ -7976,6 +8021,15 @@ meta_window_allows_resize (MetaWindow *window)
 }
 
 void
+meta_window_get_minimum_size (MetaWindow   *window,
```
Comment (cosimoc) [inline src/core/window.c]: Should this not be called `get_minimum_size_hints()`?

File: src/core/constraints.c [hunk 3]
```diff
@@ -994,6 +1000,59 @@ constrain_size_limits (MetaWindow         *window,
...
+    return TRUE;
+
+  window->currently_allocating_rect = info->current;
+  meta_window_start_geometry_allocation (window);
+
+  too_small = !meta_rectangle_could_fit_rect (&info->current,
+                                              &window->currently_allocating_rect);
+
+  if (check_only || !too_small)
+    return too_small;
```
Comment (cosimoc) [inline src/core/constraints.c]: I would flip the polarity of `too_small` around, and rename it to `can_fit`

File: src/core/window.c [hunk 4]
```diff
@@ -649,6 +650,24 @@ meta_window_class_init (MetaWindowClass *klass)
...
                   NULL, NULL, NULL,
                   G_TYPE_NONE, 0);
+
+  /**
+   * MetaWindow::geometry-allocate:
+   * @window: A #MetaWindow
+   *
+   * This is emitted when a window is about to be resized,
+   * allowing connectors to override the size of that window if
+   * required. Use meta_window_expand_minimum_size to do that.
```
Comment (cosimoc) [inline src/core/window.c]: This comment is outdated and mentions a method name that does not exist anymore.

File: src/core/window.c [hunk 5]
```diff
@@ -3652,6 +3671,32 @@ meta_window_update_monitor (MetaWindow *window,
...
+
+void
+meta_window_expand_allocated_geometry (MetaWindow *window,
+                                       int        width,
+                                       int        height)
+{
+    MetaRectangle next = {
+        .x = window->currently_allocating_rect.x,
+        .y = window->currently_allocating_rect.y,
+        .width = (width < window->currently_allocating_rect.width ?
```
Comment (cosimoc) [inline src/core/window.c]: Use `MAX()`

File: src/core/window.c (hunk 1 above)
Comment (smspillaz) [inline src/core/window.c]: I'd like to keep this function around as it "completes" the API, similar to the way that gtk_widget_size_allocate works.

File: src/core/window.c (hunk 2 above)
Comment (smspillaz) [inline src/core/window.c]: It could be - though in reality its not just a hint but more of a constraint.

File: src/core/constraints.c (hunk 3 above)
Comment (smspillaz) [inline src/core/constraints.c]: Actually, it looks like this check was the wrong way around as well! Surprised it works ...

**Comment by smspillaz** [comment]
Fixed all comments except the unused function and rebased the branch.

File: src/core/window.c (hunk 1 above)
Comment (cosimoc) [inline src/core/window.c]: I disagree; the API is private between mutter and the shell, and we are adding this as a downstream patch. If we are not even using it, what's the point of adding it and maintaining it through every rebase?

File: src/core/window.c [hunk 6]
```diff
@@ -649,6 +650,24 @@ meta_window_class_init (MetaWindowClass *klass)
...
+  /**
+   * MetaWindow::geometry-allocate:
+   * @window: A #MetaWindow
+   *
+   * This is emitted when a window is about to be resized,
+   * allowing connectors to override the size of that window if
+   * required. Use meta_window_expand_allocated_geometry to do that.
+   */
+  window_signals[GEOMETRY_ALLOCATE] =
+    g_signal_new("geometry-allocate",
```
Comment (cosimoc) [inline src/core/window.c]: Missing space before paren

File: src/core/window.c [hunk 7]
```diff
@@ -8008,3 +8060,12 @@ meta_window_emit_size_changed (MetaWindow *window)
 {
   g_signal_emit (window, window_signals[SIZE_CHANGED], 0);
 }
+
+void
+meta_window_start_geometry_allocation (MetaWindow *window)
+{
+  g_signal_emit(window,
```
Comment (cosimoc) [inline src/core/window.c]: Missing space before paren

File: src/core/window.c [hunk 8]
```diff
@@ -3652,6 +3671,30 @@ meta_window_update_monitor (MetaWindow *window,
...
+
+void
+meta_window_expand_allocated_geometry (MetaWindow *window,
+                                       int        width,
+                                       int        height)
+{
+    MetaRectangle next = {
+        .x = window->currently_allocating_rect.x,
+        .y = window->currently_allocating_rect.y,
+        .width = MAX(window->currently_allocating_rect.width, width),
```
Comment (cosimoc) [inline src/core/window.c]: Missing space before paren

File: src/core/window.c (hunk 1 above)
Comment (smspillaz) [inline src/core/window.c]: Hmm okay. I guess if we're not concerned about upstreamability then we don't need to worry about the corresponding API.

**Comment by smspillaz** [comment]
Okay, comments addressed, pushed and rebased.

File: src/core/window.c [hunk 9]
```diff
@@ -3652,6 +3671,22 @@ meta_window_update_monitor (MetaWindow *window,
...
+void
+meta_window_expand_allocated_geometry (MetaWindow *window,
+                                       int        width,
+                                       int        height)
+{
+    MetaRectangle next = {
+        .x = window->currently_allocating_rect.x,
+        .y = window->currently_allocating_rect.y,
+        .width = MAX (window->currently_allocating_rect.width, width),
+        .height = MAX (height, window->currently_allocating_rect.height)
```
Comment (cosimoc) [inline src/core/window.c]: Could you fix these two lines to use the parameters in the same order? #ocd

Comment (smspillaz) [inline src/core/window.c]: Heh, that does look a bit odd now that I think of it. Will do :)

**Pull request merged**